contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -380,60 +459,76 @@ contract LidoOracle is ILidoOracle, AragonApp {
-            if (mask.getBit(index)) {
-                data[i++] = reportToUint256(gatheredEpochData[_epochId].reports[index]);
+        // if more than 2 kind of reports exist, choose the most frequent
+        uint256 maxi = 0;
+        for (uint256 i = 1; i < gatheredReportKinds.length; ++i) {
+            if (gatheredReportKinds[i].count > gatheredReportKinds[maxi].count) maxi = i;
+        }
+        for (i = 0; i < gatheredReportKinds.length; ++i) {
+            if (i != maxi && gatheredReportKinds[i].count == gatheredReportKinds[maxi].count) {
+                return (false, Report({beaconBalance: 0, beaconValidators: 0}));
             }
         }
-
-        assert(i == data.length);
-
-        // find mode value of this array
-        (bool isUnimodal, uint256 mode) = Algorithm.mode(data);
-        if (!isUnimodal)
-            return (false, Report({beaconBalance: 0, beaconValidators: 0}));
-
-        // unpack Report struct from uint256
-        modeReport = uint256ToReport(mode);
-
-        return (true, modeReport);
+        return (gatheredReportKinds[maxi].count >= quorum, gatheredReportKinds[maxi].report);
     }
 
     /**
      * @dev Pushes the current data point if quorum is reached
      */
-    function _tryPush(uint256 _epochId) internal {
-        (bool isQuorum, Report memory modeReport) = _getQuorumReport(_epochId);
-        if (!isQuorum)
-            return;
+    function _tryPush(uint256 epochId) internal {
+        (bool isQuorum, Report memory modeReport) = _getQuorumReport();
+        if (!isQuorum) return;
 
         // data for this frame is collected, now this frame is completed, so
-        // minReportableEpochId should be changed to first epoch from next frame
+        // reportableEpochId should be changed to first epoch from the next frame
         BeaconSpec memory beaconSpec = _getBeaconSpec();
-        MIN_REPORTABLE_EPOCH_ID_POSITION.setStorageUint256(
-            _epochId
+        emit Completed(epochId, modeReport.beaconBalance, modeReport.beaconValidators);
+        _clearReportingAndAdvanceTo(
+            epochId
             .div(beaconSpec.epochsPerFrame)
             .add(1)
-            .mul(beaconSpec.epochsPerFrame)
-        );
-
-        emit Completed(_epochId, modeReport.beaconBalance, modeReport.beaconValidators);
+            .mul(beaconSpec.epochsPerFrame));
 
+        // report to the Lido and collect stats
         ILido lido = getLido();
-        if (address(0) != address(lido))
-            lido.pushBeacon(modeReport.beaconValidators, modeReport.beaconBalance);
+        uint256 prevTotalPooledEther = ISTETH(lido).totalSupply();
+        lido.pushBeacon(modeReport.beaconValidators, modeReport.beaconBalance);
+        uint256 postTotalPooledEther = ISTETH(lido).totalSupply();
+
+        PRE_COMPLETED_TOTAL_POOLED_ETHER_POSITION.setStorageUint256(prevTotalPooledEther);
+        POST_COMPLETED_TOTAL_POOLED_ETHER_POSITION.setStorageUint256(postTotalPooledEther);
```

Comment:
Do we need to store this value in Oracle smart contract? Because we've already have this value in Lido. What is the motivation behind this?

Reply:
Yes, but this is the snapshot of the Lido's `totalPooledEther` for the time of `lido.pushBeacon()` tx. In any other time it may change (e.g. because of stETH purchase).